crucible/src/Lang/Crucible/Simulator/SymSequence.hs

```diff
@@ -0,0 +1,507 @@
+   f _loop SymSequenceNil = pure (SeqTail Unassigned)
+   f _loop (SymSequenceCons _ _v tl) = pure (SeqTail (justPartExpr sym tl))
+   f loop (SymSequenceMerge _ p xs ys) =
+     do tx <- getSeqTail <$> loop xs
+        ty <- getSeqTail <$> loop ys
+        SeqTail <$> mergePartial sym f' p tx ty
+   f loop (SymSequenceAppend _ xs ys) =
+     loop xs >>= \case
+       SeqTail Unassigned -> loop ys
+       SeqTail (PE px tx)
+         | Just True <- asConstantPred px ->
+             do t <- appendSymSequence sym tx ys
+                pure (SeqTail (PE px t))
+
+         | otherwise ->
+             loop ys >>= \case
+               SeqTail Unassigned -> pure (SeqTail (PE px tx))
+               SeqTail (PE py ty) ->
+                 do p <- orPred sym px py
+                    t <- appendSymSequence sym tx ys
+                    SeqTail <$> runPartialT sym p (f' px t ty)
+
+
+{-# SPECIALIZE
+  traverseSymSequence ::
+  sym ->
+  (a -> IO b) ->
+  SymSequence sym a ->
+  IO (SymSequence sym b)
+ #-}
+
+-- | Visit every element in the given symbolic sequence,
+--   applying the given action, and constructing a new
+--   sequence. The traversal is memoized, so any given
+--   subsequence will be visited at most once.
+traverseSymSequence :: forall m sym a b.
+  MonadIO m =>
+  sym ->
+  (a -> m b) ->
+  SymSequence sym a ->
+  m (SymSequence sym b)
+traverseSymSequence sym f = \s -> getConst <$> evalWithFreshCache fn s
+  where
+   fn :: (SymSequence sym a -> m (Const (SymSequence sym b) a)) ->
+         (SymSequence sym a -> m (Const (SymSequence sym b) a))
+   fn _loop SymSequenceNil = pure (Const SymSequenceNil)
+   fn loop (SymSequenceCons _ v tl) =
+     do v'  <- f v
+        tl' <- getConst <$> loop tl
+        liftIO (Const <$> consSymSequence sym v' tl')
+   fn loop (SymSequenceAppend _ xs ys) =
+     do xs' <- getConst <$> loop xs
+        ys' <- getConst <$> loop ys
+        liftIO (Const <$> appendSymSequence sym xs' ys')
+   fn loop (SymSequenceMerge _ p xs ys) =
+     do xs' <- getConst <$> loop xs
+        ys' <- getConst <$> loop ys
+        liftIO (Const <$> muxSymSequence sym p xs' ys')
+
+
+-- | Using the given evaluation function for booleans, and an evaluation
+--   function for values, compute a concrete sequence corresponding
+--   to the given symbolic sequence.
+concreteizeSymSequence ::
+  (Pred sym -> IO Bool) {- ^ evaluation for booleans -} ->
+  (a -> IO b) {- ^ evaluation for values -} ->
+  SymSequence sym a -> IO [b]
+concreteizeSymSequence conc eval = loop
+  where
+    loop SymSequenceNil = pure []
+    loop (SymSequenceCons _ v tl) = (:) <$> eval v <*> loop tl
+    loop (SymSequenceAppend _ xs ys) = (++) <$> loop xs <*> loop ys
+    loop (SymSequenceMerge _ p xs ys) =
+      do b <- conc p
+         if b then loop xs else loop ys
+
+-- | Given a pretty printer for elements,
+--   print a symbolic sequence.
+prettySymSequence :: IsExpr (SymExpr sym) =>
```

Comment:
Would it be more convenient to also have a `Pretty` instance with `pretty = prettySymSequence pretty`?